### ISSUE TITLE
Extend Livelihoods Modelling for wider scope of Change Scenarios

### ISSUE THREAD

**Issue by daiClegg** [dbrownukk/EFD_v2#440]
**Current Capability**

A Change Scenario in IDAPs Livelihoods Modelling (OHEA & OIHM) defines one or more changes in yield and price for crops, which are then used to recalculate:
- Cash Income, 
- Food Income, 
- DI 
- DI with Standard of Living Costs (2 levels for OHEA, one for OIHM) 

These are then compared with the original values of these calculations as derived from the  actual recorded study data to show the impact of the Change Scenario.

**Proposed New Capability**

The modelling would have greater value and be applicable in more real-world situations if the Change Scenario were able to include other variations in the data set:
These are:

- *Livestock:*  %age change in no. of units available to sell; %age change in price per unit
- *Livestock Products:* %age change in no. of units produced; %age change in price per unit
- *Employment:* %age change in no. of units worked; %age change in payment per unit
- *Employment:* New employment no of units worked and payment per unit
- *Transfers:* %age change amount of transfer (specific source) 
- *Transfers:* New transfer amount & source  
- *Wild Foods:* %age change in no. of units produced; %age change in price per unit

**Comment by daiClegg** [comment]
Also discussed how changes in access to markets might be modelled, but no specific requirement was identified. The above definition assumes it would be represented in changes in price

**Comment by DrJohnSeaman** [comment]
Sorry about the delay. I have been wihout the internet

Current Capability

A Change Scenario in IDAPs Livelihoods Modelling (OHEA & OIHM) defines one or more changes in yield and price for crops, which are then used to recalculate:

    Cash Income,
    Food Income,
    DI
    DI with Standard of Living Costs (2 levels for OHEA, one for OIHM)
_and %change and price change for livestock sales, livestock product sales, employment(%change in avialability of work), % change in transfers and transfer prices (for example the sale of food aid) and % chnage and price change of wild foods (which in some places are an improtant food income source)_
_and the model is run within the parameters set by the user ie food prioes, Kcal requirement (currently set at 2100kcal but could be changeable) and the set 'coping strategies' e.g.livestock sales.__

These are then compared with the original values of these calculations as derived from the actual recorded study data to show the impact of the Change Scenario.

Proposed New Capability

The modelling would have greater value and be applicable in more real-world situations if the Change Scenario were able to include other variations in the data set:
These are:

    Livestock: %age change in no. of units available to sell; %age change in price per unit.  _
_.I thought that this was already there - as an income source it can be changed (and the price). Mark Lawrence's algoritjm  specifies maximum offtake deals with sales as a 'coping startegy'_ 
    Livestock Products: %age change in no. of units produced; %age change and chnage in price per unit - _this should be included_ 
    Employment: %age change in no. of units worked; %age change in payment per unit. _Yes, see above_
    Employment: New employment no of units worked and payment per unit
_I am not sur eI undertand this - if it means add a new type of employment as a coping strategy then I wouldnt object but I cant think when it would be used._
    Transfers: %age change amount of transfer (specific source)
    Transfers: New transfer amount & source
    Wild Foods: %age change in no. of units produced; %age change in price per  unit - _these should be included_

Also discussed how changes in access to markets might be modelled, but no specific requirement was identified. The above definition assumes it would be represented in changes in price...correct.

**Comment by celiapetty** [comment]
Re new coping employment: James explained on the call that this would represent a change from the baseline and  would need to be  reflected in a baseline update. James, could you elaborate on how a 'new' coping  employment type (ie something not shown in the baseline) would be dealt with in a study?

**Comment by daiClegg** [comment]
Interesting point John makes above - that food aid (Transfer type-  Food, Source type - Official, Source - xxx NGO) can of course, be sold.
The option to sell more would be an expandability rule; this issue is about change scenarios, so i'll take that out of scope for this issue and log it separately.

**Comment by celiapetty** [comment]
This enhancement would be very helpful for the NIMFRU /LIMB work (seasonal changes in livestock prices after flooding-many more on the market to make up lost crop food and cash income). Also changes in employment availability ( Agric lab dropped to 0 in James' 2019 Katakwi monitoring field visit ) .  Do we need to look at budgets and timelines. I know there are many competing priorities ?

**Comment by daiClegg** [comment]
This is a non-trivial effort. 
I think we need:

1. A ballpark estimate (potentially broken down by the different dimensions of expansion (Livestock, Employment, etc) 
2. Prioritisation of the multiple dimensions (if it's all-or-nothing, fine, but if there are lower value elements we can assess cost with or without them)
3. Details specification for the extensions  - no point nailing all the detail until we have the done 2 previous talks. 
4. Detail specs  might change costs, which might affect priorities, so we should finalise the work package before we start 

David & i will start on 1.
@celiapetty  - can you take a look at 2, please?

**Comment by celiapetty** [comment]
Priorities: Both Livestock and livestock products  and employment are first priorities. In different circumstances either could be most significant.
I would definitely start with these two, then look at budgets and have a wider discussion on use cases likely to come up in new projects eg Bangladesh.

**Comment by daiClegg** [comment]
Thx. Discussing point 1 tomorrow.

**Comment by daiClegg** [comment]
Let's see what is supported in the latest change

**Comment by dbrownukk** [comment]
Change Scenario modelling extended as per https://github.com/dbrownukk/EFD_v2/files/6542581/ModellingExtensionsCS.ER3.xlsx

Added Wild Foods, Livestock Products, Livestock Sales, Employment (Cash and Food), Transfers (Cash and Food) 

Note that a Transfer Price Yield Variation now has an additional field for the transferred food type where Transfer Type is Food. This is only available when Transfer is of type Food and then used in Modelling Report

Deployed to IDAPS_Livelihoods_TESTING

**Comment by daiClegg** [comment]
see also #443 for training event requirement (December 2021), but need it in production well in advance for preparing training materials based in it.

**Comment by daiClegg** [comment]
@DrJohnSeaman @celiapetty 

I'm testing the impact of changing "yield" and "price" of all the different resources (Crops, Employment, LS Sales, LS Products, Transfers, Wild Foods).

Crops are well established in the software.
And Wild Foods now behave exactly the same way.
I'm guessing for LS Products it is legitimate to vary yield (e.g smaller animals will result in less meat) as well as price. So these too will behave just like Crops

For Employment, though the meanings are subtly different, we might model a scenario in which there is only 60% of Boda Boda Employment available during lockdown, and prices fall because more drivers are chasing less custom. So "Yield" is actually taken to mean the availability of an Employment type; Price is the Payment - the Price of the Labour, if we're going to go all Marxian about it!
Is that a useful feature?

But does it make sense to vary the"yield" of Livestock Sales?
For example, modelling a dry year we might say only 80% of the new calves survive and the ones that do make it to market are scrawny and only command 70% of the normal price
Should we support that option? Or should we say you can only vary price for live sales?

And what does a price/yield mean for a transfer?
I can imagine that we might want to model the impact of a pension cut, but how would we know which transfer to apply that to? We do not capture the reason for a cash transfer, only optionally the source (Kin/Government/NGO etc). 
But even if we did capture and then specify the Transfers affected, the concept of Yield, i think, is irrelevant, we're just saying the amount (Price) of the Transfer changes. 

Note that in a couple of instances here we're overloading the concepts of Yield & Price. How that is manages in the UI, so the user can understand what is going on, is a different matter than what we do with the numbers and indeed whether we even allow the ser to change them on certain types of Variation,

**Comment by daiClegg** [comment]
Not removed Testing label pending resolution of the issues above, 'cos i can continue to test the uncontroversial changes, while the more arcane are resolved,

**Comment by daiClegg** [comment]
Testing Notes:

I can set up Variations for all resource types, 
And i can run Change Scenario reports successfully for them all 

*I need to check that the calculations are correct for the different types*
*i'm not sure that Livestock Sales "Yield" is being taken into account, but that might be the correct implementation , depending on the issues above.

**Comment by daiClegg** [comment]
@celiapetty it would be helpful to have some feedback here, or i could just complete my testing of what we have and then promote it to the QA instance so you can test and see if it makes real-world sense. That might be easier than having a lengthy discussion of what might be.
I could do a quick demo session on Zoom to show what we have.

**Comment by celiapetty** [comment]
Sounds as if a quick Zoom demo would be a good idea. When would be a good
time for you? I  have meetings 11.30-12.30   and 2.30-3.30 tomorrow. Wed am
free  11.20-12.30 and 1.30-3.30

**Comment by DrJohnSeaman** [comment]
wednesday is best for me John

**Comment by daiClegg** [comment]
Instance: Testing
App: Modelling/OIHM
Change Scenario: Test Extent of Variations
Study: Katakwi - NIMFRU - AIHM 2018

Cash
---

HH3: 
- Crop change **correct**
- Employment Change **correct**
However, the assumption appears to be that because the availability of the employment in question was 80% )(Yield) the 2 family members in this employment, worked 80% of a year each.
Is that valid?

HH9:
- Livestock (Products & Live Sales) **correct**
However, the calculation appears to assume that if the availability of Animals to sell is 80% that still rounds up to 1 for this family; the Price change is applied correctly.
Is that valid?

tbc...

**Comment by celiapetty** [comment]
HH3
Here, we only know that in the course of a year, the household earned
216,000 in agric labour (not the number of days or the rate per day. ) The
assumption that they have 20% less work so over a year they would work 20%
fewer days is reasonable. Is that the question? Future training will
emphasise the importance of specifying the number of days worked and rate
per day (or number of months and rate per month etc).

HH9
Cow - 1 the rule shouldn't allow this to be sold, as it would count as a
distress sale.
Chickens- 6 Non sold in the study year, but they could sell the permitted
number within the rule, I think? For this function to work (ie sell where
there is no initial sale price) we will need to insist in training that a
value is given for all livestock. Query: will it be possible for users to
override this price and insert a realistic price based on current market
trends?
If I haven't interpreted the question correctly could you let me know,
please? We can arrange a call to clarify if needed. Thanks.

**Comment by daiClegg** [comment]
Instance: Testing
App: Modelling/OIHM
Change Scenario: Test Extent of Variations
Study: Katakwi - NIMFRU - AIHM 2018

Cash (continued)
---


HH12:

Transfer  change **not as expected**
The transfer price/yield change is for 'other goods' - ie not a cash or food transfer.
But this HH only has Food transfers
so should have had no impact

HH40 Wild Food Change:

- Wold Food change **correct**

**Comment by daiClegg** [comment]
I have not tested Livestock Cash income from a Livestock Product sale, because i chose one that is always consumed - never sold.
Here is the annotated data extract i have used for testing so far ) notes in column AA 

[ChangeNIMFRUTue Jul 27 20_51_57 UTC 2021.xlsx](https://github.com/dbrownukk/EFD_v2/files/6896625/ChangeNIMFRUTue.Jul.27.20_51_57.UTC.2021.xlsx)

tbc...

**Comment by daiClegg** [comment]
Instance: Testing
App: Modelling/OIHM
Change Scenario: Further Change Scenario Extension Testing
Study: Katakwi - NIMFRU - AIHM 2018

Livestock (Livestock Products)
---
HH5:

- Livestock Change is **correct**

The data extract, annotated, that i used for this test is attached

[ChangeNIMFRUWed Jul 28 21_48_20 UTC 2021.xlsx](https://github.com/dbrownukk/EFD_v2/files/6896675/ChangeNIMFRUWed.Jul.28.21_48_20.UTC.2021.xlsx)

**Comment by daiClegg** [comment]
I still have more tests to do, but this broadly feature complete.

I will run through the equivalent changes for Food Income before closing it.
Engineering need to discuss the Transfers calculation

**Comment by daiClegg** [comment]
@celiapetty, re Your comment above:
> HH9
Cow - 1 the rule shouldn't allow this to be sold, as it would count as a
distress sale.
Chickens- 6 Non sold in the study year, but they could sell the permitted
number within the rule, I think? For this function to work (ie sell where
there is no initial sale price) we will need to insist in training that a
value is given for all livestock. Query: will it be possible for users to
override this price and insert a realistic price based on current market
trends?
If I haven't interpreted the question correctly could you let me know,
please? We can arrange a call to clarify if needed. Thanks.

Aren't you discussing expandability Rules? This is a test of a Change Scenario in which only 80% of the reported Cows would be available (e.g loss of animals due to drought) and they would command only 80% of the reported price at market (ie the animals brought to market are sub-standard).
My question is what should happen, under that change scenario, to a HH that actually sold 1 animal?

**Comment by daiClegg** [comment]
Instance: Testing
App: Modelling/OIHM
Change Scenario: Test Extent of Variations
Study: Katakwi - NIMFRU - AIHM 2018

Cash
---
**HH3:**

- Crop change **correct**
- Employment Change **not as expected**
In my example, the availability of the employment type shrank to 80% and the pay shrank to 80%
Unlike in all the other resource types, if the employment is paid in food, then the rate of pay **does** need to be taken into account, i think. 
The calculation only takes cognisance of the availability (yield), 
- Livestock (Products & Live Sales) **correct**
- Wild Food **correct**

I haven't tested Transfers yet (no relevant data in my test)

**Comment by daiClegg** [comment]
I'm satisfied that this is **feature complete**

There is one known anomaly (Cash impact of Transfers - to be logged as a separate issue.
There are some methodological questions that i will resolve with Celia & will log as separate issues if necessary.

**Comment by daiClegg** [comment]
Annotated spreadsheet used for the Food Income change tests attached here:
[ChangeNIMFRUTue Jul 27 20_51_57 UTC 2021.xlsx](https://github.com/dbrownukk/EFD_v2/files/6903843/ChangeNIMFRUTue.Jul.27.20_51_57.UTC.2021.xlsx)

**Comment by celiapetty** [comment]
Dai
Was this your most recent github post? I lost track of a message in sending
it from Spam to my inbox - I think it was probably the one? (Google still
sends all github messages to spam, although I have you and David in my
contacts list.)
 btw do you have any comments on the latest from Datakind?
Will has just had his sabbatical confirmed by Deloittes so will be joining
Walker, probably in Sept. This move will also give him time to commit  the
1- 1/2 days to Datakind during the 'discovery' period.
Celia

**Comment by daiClegg** [comment]
Just click the ‘view it in GitHub’ link at the bottom of the email.  It then shows you the whole thread in a browser tab.

D
